### PR TITLE
Add spec + proof for poly_basemul_montgomery_cached()

### PR DESCRIPTION
**Pull request by hanno-becker**
[//]: # (SPDX-License-Identifier: CC-BY-4.0)

### CODE REVIEW

**Comment by hanno-becker** [comment]
@rod-chapman Do you have insights here?

It looks like some disjointness reasoning isn't working as expected in
```c
  for (i = 0; i < MLKEM_N / 4; i++)
      // clang-format off
    ASSIGNS(i, OBJECT_WHOLE(r))
    INVARIANT(i >= 0 && i <= MLKEM_N / 4)
    // Add this, ultimately
    // INVARIANT(ARRAY_IN_BOUNDS(int, k, 0, (4 * i - 3), r->coeffs, -3 * HALF_Q, 3 * HALF_Q))
    DECREASES(MLKEM_N / 4 - i)
    // clang-format on
  {
    basemul_cached(&r->coeffs[4 * i], &a->coeffs[4 * i], &b->coeffs[4 * i],
                   b_cache->coeffs[2 * i]);
    ASSERT(r->coeffs[4*i + 0] >= - 3 * HALF_Q + 1, ""); // OK
    ASSERT(r->coeffs[4*i + 0] <=   3 * HALF_Q - 1, ""); // OK
    ASSERT(r->coeffs[4*i + 1] >= - 3 * HALF_Q + 1, ""); // OK
    ASSERT(r->coeffs[4*i + 1] <=   3 * HALF_Q - 1, ""); // OK

    basemul_cached(&r->coeffs[4 * i + 2], &a->coeffs[4 * i + 2],
                   &b->coeffs[4 * i + 2], b_cache->coeffs[2 * i + 1]);
    ASSERT(r->coeffs[4*i + 2] >= - 3 * HALF_Q + 1, ""); // OK
    ASSERT(r->coeffs[4*i + 2] <=   3 * HALF_Q - 1, ""); // OK
    ASSERT(r->coeffs[4*i + 3] >= - 3 * HALF_Q + 1, ""); // OK
    ASSERT(r->coeffs[4*i + 3] <=   3 * HALF_Q - 1, ""); // OK

    // This no longer succeeds!
    // Is something wrong with the spec of basemul_cached() that
    // prevents disjointness reasoning between r[4i:4i+1] and r[4i+2:4i+3]?
    // ASSERT(r->coeffs[4*i + 0] >= - 3 * HALF_Q + 1, "");
  }
```

**Comment by hanno-becker** [comment]
Interestingly, when temporarily removing the `ASSIGNS(OBJECT_WHOLE(r))` clause from
```c
void basemul_cached(int16_t r[2], const int16_t a[2], const int16_t b[2],
                    int16_t b_cached)
    // clang-format off
REQUIRES(r != NULL && IS_FRESH(r, 2 * sizeof(int16_t)))
REQUIRES(a != NULL && IS_FRESH(a, 2 * sizeof(int16_t)))
REQUIRES(b != NULL && IS_FRESH(b, 2 * sizeof(int16_t)))
REQUIRES(ARRAY_IN_BOUNDS(int, k, 0, 1, a, -(MLKEM_Q - 1), (MLKEM_Q - 1)))
ASSIGNS(OBJECT_WHOLE(r))
ENSURES(ARRAY_IN_BOUNDS(int, k, 0, 1, r, -3 * HALF_Q + 1, 3 * HALF_Q - 1));
// clang-format on
```
it works, hinting that indeed something's off with disjointness reasoning here.

@rod-chapman Do you know how memory is modelled in CBMC? If `OBJECT_WHOLE(r)` resolves to the entire _array_ that `r` points _into_, then it would make sense that things fail. But that would make `OBJECT_WHOLE(...)` very hard to use in specs.

**Comment by hanno-becker** [comment]
Indeed, with 
```c
ASSIGNS(__CPROVER_object_upto(r, 2 * sizeof(int16_t)))
```
it seems to work...

File: mlkem/poly.h [hunk 1]
```diff
@@ -238,8 +239,39 @@ void poly_getnoise_eta1122_4x(poly *r0, poly *r1, poly *r2, poly *r3,
...
+ * Name:        poly_basemul_montgomery_cached
+ *
+ * Description: Multiplication of two polynomials in NTT domain,
+ *              using mulcache for second operand.
+ *
+ *              Bounds:
+ *              - a is assumed to be coefficient-wise < q in absolute value.
+ *              - b is assumed to be the output of a forward NTT and
+ *                thus coefficient-wise bound by C
+ *              - b_cache is assumed to be coefficient-wise bound by D.
```
Comment (mkannwischer) [inline mlkem/poly.h:252]: I am confused by these bounds. `basemul_cached` says b and b_cache can be arbitrary int16 values. Should these bounds be relaxed here?
If I do understand correctly, then the b is bouned by `NTT_BOUND` (`8 * MLKEM_Q`) and `b_cache` does not have a bound that is guaranteed (although it likely has the same bounds as fqmul)?

**Review by mkannwischer — APPROVED** [review]
The CBMC bounds for `poly_basemul_montgomery_cached` look good.
The bounds in the comments are confusing. Please adjust those before you merge.

File: mlkem/poly.h (hunk 1 above)
Comment (hanno-becker) [reply mlkem/poly.h]: Good spot, removed this part of the doc to make it uniform with the formal spec

**Pull request merged**